REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [3]:
import pandas as pd

In [4]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [5]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [6]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

<font color='Red'>There are 6315 Records</font>

In [7]:
len(bus)

6315

#### Question 2b: How many unique business IDs are there?  

<font color='Red'>There are 6315 Unique Business ID's</font>

In [8]:
len(bus['business_id'].unique())

6315

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

<font color='Red'>The 5 most common business names are: Starbucks Coffee:72, Peet's Coffee & Tea:24, McDonalds:12, San Francisco Soup Company:11, Walgreens:11</font>

In [9]:
bus['name'].value_counts().to_frame().head(5)

name
STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

<font color='Red'>Zip codes are stored as strings. This makes sense to me, because they are an identifier and should be used for any mathematical operations. This is also good to know for string manipulation purposes and also merging with other dataframes</font>

In [10]:
type(bus['postal_code'][0])

str

#### Question 3b

What are the unique values of postal_code?

<font color='Red'>There are 46 unique values. The Unique values are shown below.</font>

In [11]:
bus['postal_code'].value_counts().to_frame()

postal_code
94110              585
94103              558
94102              458
94107              451
94133              417
94109              383
94111              282
94122              270
94118              243
94115              238
94114              230
94105              226
94108              224
94117              197
94112              190
94123              169
94124              164
94121              161
94132              134
94104              133
94116               93
94134               74
94127               71
94131               48
94158               17
94130                8
94143                5
CA                   3
Ca                   3
94101                2
94013                2
94129                2
94014                2
941033148            1
94602                1
94621                1
94188                1
00000                1
94609                1
95105                1
941                  1
94120                1
941102019            1
94545                1
94066                1
92672                1

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

<font color='Red'>There are 5999 businesses in the new dataframe</font>

In [12]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [13]:
bus_valid = bus[bus['postal_code'].isin(validZip)]

In [14]:
len(bus_valid)

5999

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

<font color='Red'>There are 2483 records with missing longitude values</font>

In [15]:
bus_valid[pd.isnull(bus_valid['longitude'])]

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

In [16]:
sum(pd.isnull(bus_valid['longitude']))

2483

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [17]:
#There's gotta be an easier way - pls excuse my OVERCOMPLICATED CODE

#initialize dataframe with postal codes
postal_code = list(bus_valid['postal_code'].value_counts().to_frame().index)
postaldf = pd.DataFrame(postal_code)
postaldf = postaldf.rename(columns={0:'postal_code'})


#how many null/not null values in each postal code?
null_counts = []
not_null_counts = []
for code in postal_code:
    zipdf = bus_valid[bus_valid['postal_code'] == code]
    null_counts.append(sum(pd.isnull(zipdf['longitude'])))
    not_null_counts.append(sum(pd.notnull(zipdf['longitude'])))
postaldf['null_lon'] = null_counts
postaldf['not_null_lon'] = not_null_counts

postaldf.head()

postal_code  null_lon  not_null_lon
0       94110       251           334
1       94103       252           306
2       94102       196           262
3       94107       247           204
4       94133       135           282

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


<font color='Red'>Zip code 94107 has the most null with 0.55 fraction null</font>

In [18]:
postaldf['missing_lon_frac'] = postaldf['null_lon'] / (postaldf['null_lon'] + postaldf['not_null_lon'])
postaldf_frac_sorted = postaldf.sort_values('missing_lon_frac', ascending=False)
postaldf_frac_sorted.head()

postal_code  null_lon  not_null_lon  missing_lon_frac
3        94107       247           204          0.547672
16       94124        88            76          0.536585
18       94132        65            69          0.485075
10       94114       108           122          0.469565
7        94122       123           147          0.455556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

<font color='Red'>Since there are more inspections than businesses, there is more than one inspection record for each business</font>

In [19]:
print(len(bus))
print(len(ins))

6315
15430


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

<font color='Red'>type takes on "routine" and "complaint". "complaint" has only 1 record while "routine" has 15429 records. I have eliminated complaints by subsetting only routine inspections.</font>

In [20]:
ins['type'].value_counts().to_frame()

type
routine    15429
complaint      1

In [21]:
ins2 = ins[ins['type'] == 'routine']
ins2

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15429 rows x 4 columns]

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [22]:
ins2['year'] = ins2['date']
year_only = lambda x: str(x)[0:4]
ins2['year'] = ins2['year'].apply(year_only)
ins2

C:\Users\wangj\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\wangj\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15429 rows x 5 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

<font color='Red'>2013 only has 38 records whereas 2015-2016 has thousands of records. ins3 is created without records of 2013.</font>

In [23]:
ins2['year'].value_counts().to_frame()

year
2016  5839
2014  5629
2015  3923
2013    38

In [24]:
ins3 = ins2[ins2['year'] != '2013']
ins3

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15391 rows x 5 columns]

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [41]:
ins3_2016 = ins3[ins3['year']=='2016']
Ratings_2016 = pd.merge(ins3_2016, bus_valid, how="inner", left_on ='business_id', right_on ='business_id')

In [53]:
Ratings_2016.head(7)

business_id  score      date     type  year  \
0           10     82  20160503  routine  2016   
1           19     94  20160513  routine  2016   
2           24     98  20161005  routine  2016   
3           24     96  20160311  routine  2016   
4           31     84  20160816  routine  2016   
5           45     84  20160614  routine  2016   
6           45     78  20160104  routine  2016   

                                 name                        address  \
0                    TIRAMISU KITCHEN                  033 BELDEN PL   
1               NRGIZE LIFESTYLE CAFE   1200 VAN NESS AVE, 3RD FLOOR   
2  OMNI S.F. HOTEL - 2ND FLOOR PANTRY  500 CALIFORNIA ST, 2ND  FLOOR   
3  OMNI S.F. HOTEL - 2ND FLOOR PANTRY  500 CALIFORNIA ST, 2ND  FLOOR   
4      NORMAN'S ICE CREAM AND FREEZES           2801 LEAVENWORTH ST    
5                 CHARLIE'S DELI CAFE                3202 FOLSOM ST    
6                 CHARLIE'S DELI CAFE                3202 FOLSOM ST    

            city state postal_code   latitude   longitude  phone_number  
0  San Francisco    CA       94104  37.791116 -122.403816  +14154217044  
1  San Francisco    CA       94109  37.786848 -122.421547  +14157763262  
2  San Francisco    CA       94104  37.792888 -122.403135  +14156779494  
3  San Francisco    CA       94104  37.792888 -122.403135  +14156779494  
4  San Francisco    CA       94133  37.807155 -122.419004           NaN  
5  San Francisco    CA       94110  37.747114 -122.413641  +14156415051  
6  San Francisco    CA       94110  37.747114 -122.413641  +14156415051

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [60]:
Ratings_2016_sorted = Ratings_2016.sort_values('score')
Ratings_2016_sorted.head(20)[['name', 'address', 'score']]

name              address  score
3842      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD      52
3828                   UNCLE CAFE       65 WAVERLY PL      55
4833                 CRAZY PEPPER   2257 SAN JOSE AVE      55
5536                    POKI TIME     2101 LOMBARD ST      55
4368                   GOLDEN WOK   295 B ORIZABA AVE      56
2463             HING WANG BAKERY        339 JUDAH ST      56
696                  RED A BAKERY      634 CLEMENT ST      57
4694    L & G VIETNAMESE SANDWICH         602 EDDY ST      57
3350      SAN TUNG RESTAURANT LLC      1031 IRVING ST      57
3851  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST      57
457           MA'S DIM SUM & CAFE      1315 POWELL ST      57
4190                   TOP'S CAFE  2460 SAN BRUNO AVE      57
5066    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST      58
3363                        ITHAI         720 POST ST      59
523           JOANIE'S DINER INC.   1329 COLUMBUS AVE      59
2884                  JADE GARDEN      3608 BALBOA ST      59
5250         PAK NAM THAI CUISINE       655 LARKIN ST      59
3930                 WOLFES LUNCH        1220 16TH ST      60
4752          VALLARTA'S TACO BAR     2022 MISSION ST      60
3815             HONG KONG LOUNGE     5322 GEARY BLVD      60

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!

NICE!